core/object/class_db.cpp

```diff
@@ -70,6 +70,138 @@ HashMap<StringName, ClassDB::ClassInfo> ClassDB::classes;
 HashMap<StringName, StringName> ClassDB::resource_base_extensions;
 HashMap<StringName, StringName> ClassDB::compat_classes;
 
+#ifdef TOOLS_ENABLED
+HashMap<StringName, ObjectGDExtension> ClassDB::placeholder_extensions;
+
+class PlaceholderExtensionInstance {
+	StringName class_name;
+	HashMap<StringName, Variant> properties;
+
+public:
+	PlaceholderExtensionInstance(const StringName &p_class_name) {
+		class_name = p_class_name;
+	}
+
+	~PlaceholderExtensionInstance() {}
+
+	void set(const StringName &p_name, const Variant &p_value) {
+		bool is_default_valid = false;
+		Variant default_value = ClassDB::class_get_default_property_value(class_name, p_name, &is_default_valid);
+
+		// If there's a default value, then we know it's a valid property.
+		if (is_default_valid) {
+			properties[p_name] = p_value;
+		}
+	}
+
+	Variant get(const StringName &p_name) {
+		const Variant *value = properties.getptr(p_name);
+		Variant ret;
+
+		if (value) {
+			ret = *value;
+		} else {
+			bool is_default_valid = false;
+			Variant default_value = ClassDB::class_get_default_property_value(class_name, p_name, &is_default_valid);
+			if (is_default_valid) {
+				ret = default_value;
+			}
+		}
+
+		return ret;
+	}
+
+	static GDExtensionBool placeholder_instance_set(GDExtensionClassInstancePtr p_instance, GDExtensionConstStringNamePtr p_name, GDExtensionConstVariantPtr p_value) {
+		PlaceholderExtensionInstance *self = (PlaceholderExtensionInstance *)p_instance;
+		StringName name = *(StringName *)p_name;
```

Comment:
This would probably not need to be copied (even if `StringName` copies are quite cheap)?

Reply:
In my latest push, I've change this (and a couple other places) to use const references to avoid the copy (at least until absolutely necessary).